pkg/volume/csi/nodeinfomanager/nodeinfomanager.go

```diff
@@ -86,13 +92,124 @@ type Interface interface {
 // NewNodeInfoManager initializes nodeInfoManager
 func NewNodeInfoManager(
 	nodeName types.NodeName,
-	volumeHost volume.VolumeHost) Interface {
+	volumeHost volume.VolumeHost,
+	migratedDrivers map[string](func() bool)) Interface {
 	return &nodeInfoManager{
-		nodeName:   nodeName,
-		volumeHost: volumeHost,
+		nodeName:        nodeName,
+		volumeHost:      volumeHost,
+		migratedDrivers: migratedDrivers,
 	}
 }
 
+// InitializeMigratedDrivers will create CSINodeInfo objects for all drivers that are to
+// be migrated. It will initialize most fields with dummy values. The CSIDriverInfo object
+// will be initialized with DriverName, ProxyInTreeVolumePlugin, and Available.
+func (nim *nodeInfoManager) InitializeMigratedDrivers() error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CSINodeInfo) ||
+		!utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) {
+		klog.Warningf("InitializeMigratedDrivers called but CSINodeInfo or CSIMigration is disabled")
+		return nil
+	}
+
+	var updateErrs []error
+
+	csiKubeClient := nim.volumeHost.GetCSIClient()
+	if csiKubeClient == nil {
+		return fmt.Errorf("error getting CSI client")
+	}
+
+	err := wait.ExponentialBackoff(updateBackoff, func() (bool, error) {
+		nodeInfo, err := csiKubeClient.CsiV1alpha1().CSINodeInfos().Get(string(nim.nodeName), metav1.GetOptions{})
+		if nodeInfo == nil || errors.IsNotFound(err) {
+			nodeInfo, err = nim.CreateCSINodeInfo()
+		}
+		if err != nil {
+			updateErrs = append(updateErrs, err)
```

Comment:
this errors list will grow without bound if the create call perpetually fails, right?

Reply:
if the create call returns a NotFound error, that means the API is not available. do we want to keep retrying in that case?

Reply:
yes, it could grow without bound (assuming we retry indefinitely like you suggested in another comment). What is the canonical way of dealing with this? It doesn't seem like we can have it both ways (have an error array with useful information, and have it not grow forever) unless we have some extra logic maybe to ignore duplicate errors?

We want to keep retrying forever, Kubelet should not/can not start without this.

Reply:
the call site that calls this also retries... retries inside retries get really tricky to reason about, and we probably want to surface the error via node status. can we drop the retry in InitializeMigratedDrivers(), return errors if they are encountered, and count on the caller to re-call InitializeMigratedDrivers()?